File: scenarios/web-console/ui/src/test-utils/modeltest/traces.js
export const validateTraces = (traces, transition) => {
    const errors = [];
    traces.forEach((trace, traceIndex) => {
        const traceName = trace.name ?? `trace ${traceIndex}`;
        let state = trace.initial;
        trace.steps.forEach((step, stepIndex) => {
            const stepName = step.name ?? `step ${stepIndex}`;
            try {
                const got = transition(state, step.event);
                if (step.wantError) {
                    errors.push(`${traceName}/${stepName}: expected error, got success`);
                }
                if (got !== step.want) {
                    errors.push(`${traceName}/${stepName}: got state ${String(got)}, want ${String(step.want)}`);
                }
                state = got;
            }
            catch (error) {
                if (!step.wantError) {
                    errors.push(`${traceName}/${stepName}: unexpected error: ${error instanceof Error ? error.message : String(error)}`);
                }
                if (state !== step.want) {
                    errors.push(`${traceName}/${stepName}: got state ${String(state)}, want ${String(step.want)}`);
                }
            }
        });
    });
    return errors;
};
export const replayTraces = (traces, transition) => {
    const errors = validateTraces(traces, transition);
    if (errors.length > 0) {
        throw new Error(`trace replay mismatch:\n${errors.map((err) => `  - ${err}`).join("\n")}`);
    }
};

File: packages/proto/modeltest/traces.go
package modeltest

import "fmt"

// TraceStep describes one event in a replayed workflow trace.
type TraceStep[S comparable, E comparable] struct {
	Name    string
	Event   E
	Want    S
	WantErr bool
}

// Trace describes an ordered workflow path to replay against production code.
type Trace[S comparable, E comparable] struct {
	Name    string
	Initial S
	Steps   []TraceStep[S, E]
}

// ValidateTraces replays traces against transition and reports all mismatches.
func ValidateTraces[S comparable, E comparable](
	traces []Trace[S, E],
	transition Transition[S, E],
) []error {
	var errs []error
	if transition == nil {
		return []error{fmt.Errorf("transition must not be nil")}
	}

	for traceIndex, trace := range traces {
		traceName := trace.Name
		if traceName == "" {
			traceName = fmt.Sprintf("trace %d", traceIndex)
		}
		state := trace.Initial
		for stepIndex, step := range trace.Steps {
			stepName := step.Name
			if stepName == "" {
				stepName = fmt.Sprintf("step %d", stepIndex)
			}
			got, err := transition(state, step.Event)
			if step.WantErr {
				if err == nil {
					errs = append(errs, fmt.Errorf("%s/%s: expected error, got nil", traceName, stepName))
				}
			} else if err != nil {
				errs = append(errs, fmt.Errorf("%s/%s: unexpected error: %v", traceName, stepName, err))
			}
			if got != step.Want {
				errs = append(errs, fmt.Errorf("%s/%s: got status %v, want %v", traceName, stepName, got, step.Want))
			}
			state = got
		}
	}
	return errs
}

// ReplayTraces fails t when any trace diverges from the transition function.
func ReplayTraces[S comparable, E comparable](
	t TestingT,
	traces []Trace[S, E],
	transition Transition[S, E],
) {
	t.Helper()
	if errs := ValidateTraces(traces, transition); len(errs) > 0 {
		t.Fatalf("trace replay mismatch:\n%s", formatErrors(errs))
	}
}
